src/base/utils/version.h

```diff
@@ -62,6 +62,12 @@ namespace Utils
         {
         }
 
+        template <typename Other>
+        constexpr Version(const std::array<Other, N> &version)
+            : m_components (version)  // GCC 4.8.4 has problem with the uniform initializer here
```

Comment:
It's very suspicious. What's really going on here (looks like you're trying to create an object of one type from an object of another type)? And why is it necessary?

Reply:
This thread: https://github.com/qbittorrent/qBittorrent/pull/9113#discussion_r196503205

Reply:
Why not the following?
```c++
constexpr Version(const ComponentsArray &version)
    : m_components(version)
{
}
```
IMO, your template is illegal in general!

Reply:
Besides, it's not clear to me why we need to create an object of a different type to implement `fullVersionString()`?
I would implement `toString(bool full)` instead (and use it inside `operator QString()` as well).

Reply:
> IMO, your template is illegal in general!

Do you mind explaining that?

Reply:
>`toString(bool full)`

or even use some descriptive enum instead of bool.

Reply:
>Do you mind explaining that?

Generally `some_template<type1>` and `some_template<type2>` are different types. And so `some_template<type1>` is unlikely to have `some_template<type1>(some_template<type2> other)` constructor.
Unless I'm wrong the template above can have only one possible instantiation when `Other = T`. But it just degenerates into `constexpr Version(const ComponentsArray &version)`.

Reply:
But even if I'm wrong, my second comment (https://github.com/qbittorrent/qBittorrent/pull/9113#discussion_r197521842) is still more important.

Reply:
`std::array` has no constructors,  it uses aggregate initialisation.

Reply:
Then what really happens here: `m_components (version)`? Isn't it ` std::array<T, N>(std::array<Other, N>)`? I see no "aggregate initialisation" syntax here.

Reply:
> toString(bool full)

Inelegant, IMO, and since we are in a template class, requires generating more code while given instantiation most likely needs only one.

> or even use some descriptive enum instead of bool.

A bit better, but we already have operator QString, and it is probably used somewhere. 

Frankly speaking, I don't understand why it is needed to declare python version as  Version<T, 3,1>, why not Version <T, 3,3>?

Reply:
>  I see no "aggregate initialisation" syntax here.

Sure, there is no aggregate initialisation here.

Reply:
> we already have operator QString, and it is probably used somewhere.

As I said before:
>implement `toString(bool full)` instead (and *use it inside `operator QString()` as well*).

Reply:
>`std::array` *has no constructors*, it uses aggregate initialisation.

>Sure, there is no aggregate initialisation here. It's a call to the implicit *copy constructor*.

So, does it have constructors or not?.. You don't have to answer, they're still there! At least default and copy ones.

>It's a call to the implicit copy constructor.

Now, please reread [this](https://github.com/qbittorrent/qBittorrent/pull/9113#discussion_r197524310) comment. The `template <typename Other> constexpr Version(const std::array<Other, N> &version) : m_components(version) {}` is illegal in all cases except the one when `Other = T`.

Reply:
>Frankly speaking, I don't understand why it is needed to declare python version as Version<T, 3,1>, why not Version <T, 3,3>?

I was thinking python 2.7.0 might be declared as `2.7` (same for python 3.0 as `3` or `3.0`). Does `python -V` guarantee to return 3 digits version?

Reply:
>Does python -V guarantee to return 3 digits version?

Just tried it, python 3.0 version string returns `3.0`.
Unless someone objects, I'll keep `Version<T, 3, 1>`, there isn't much difference anyway.

Reply:
I am objecting. The only use-case for this new function I found is to produce a string that gets *parsed* back into three numbers. Thus my strong suggestion: remove the function and make a direct version comparison as I already suggested there.

Reply:
>remove the function

Alright, removing the controversial commit.

>direct version comparison as I already suggested there.

DIY, I don't care any opinions about making it "better".
I already done what I've wanted and let it be another stalled PR as the ~50 others.